### PR TITLE
Adds profiling flags, computes train metrics average.

### PR DESCRIPTION
**Pull request by andsteing**
List of small changes accumulated over the weekend:

1. Added profiling flags `--profile_steps`, `--profile_validation`, and `--profile_memory`, removed unused `--logging_dir`. Updated `--output_dir` to default to `runs/{timestamp}`.
1. Updated commands in README to store model in `runs/` subdirectory with timestamp, so they ignored by Git (`runs/` is in `.gitignore`), and multiple runs land automatically in different directories. Also setting `--tracker_project_name=$HUB_MODEL_ID`, and mention dataset for both commands to avoid mixing up different dataset runs in the same `wandb` project.
1. Changed `--num_train_epochs=11` for the `fusing/fill50k` dataset, since I noticed that models don't actually converge before this many epochs (see runs with different number of epochs here: https://wandb.ai/andsteing/controlnet-fill-circle ). Also set `--max_train_samples 100000` for the second commands, since that's what is recommended in the linked Blog article https://huggingface.co/blog/train-your-controlnet
1. Added `l2_grads` metric for logging the L2 norm of gradients.
1. The `train/step` metric, which is used as the X coordinate for other metrics, now uses the added (relative) `walltime` metric as its X coordinate. This allows to quickly read out the run time in the metric view of `wandb`.
1. Added `train/steps_per_sec` metric.
1. Bugfix computation `train_metrics`: Before this PR, the latest `train_metric` was reported, which resulted in a very high variance in e.g. the reported `loss`. Now, the average of the `train_metrics` since the last logging is reported.
1. The metric `epoch` is now fractional. This allows e.g. for plotting of metrics when experimenting with different batch sizes.

Latest runs:

- `--dataset_name=fusing/fill50k`: https://wandb.ai/andsteing/controlnet-fill-circle_bs
- `--dataset_name=multimodalart/facesyntheticsspigacaptioned`: https://wandb.ai/andsteing/controlnet-uncanny-faces

### CODE REVIEW

**Comment by sayakpaul** [comment]
@andsteing thanks for this! 

I am wondering why are changes unrelated to this PR are showing up. Could we adjust the PR so that only the relevant changes are present?

**Comment by andsteing** [comment]
d'oh ... I rebased it on the wrong branch in my cloud VM.

Fixed, please take another look.

File: examples/controlnet/README.md [hunk 1]
```diff
@@ -382,13 +388,12 @@ python3 train_controlnet_flax.py \
...
-	--max_train_samples 50 \
-	--max_train_steps 5 \
+	--max_train_samples 100000 \
 	--learning_rate=1e-5 \
-	--validation_steps=2 \
 	--train_batch_size=1 \
 	--revision="flax" \
-	--report_to="wandb"
+	--report_to="wandb" \
+  --tracker_project_name=$HUB_MODEL_ID
```
Comment (sayakpaul) [inline examples/controlnet/README.md:396]:
```suggestion
    --tracker_project_name=$HUB_MODEL_ID
```

**Review by sayakpaul — APPROVED** [review]
Neat changes.

**Review by patrickvonplaten — APPROVED** [review]
Very cool! @yiyixuxu can you double check here maybe? :-)

**Review by yiyixuxu — APPROVED** [review]
Thank you so much @andsteing 
think will use this to improve our other flax script too

**Comment by sayakpaul** [comment]
@andsteing since the PR introduces nice support for profiling, I think it'd make sense to comment on that from the example README. Could you also add a few lines there?

**Comment by andsteing** [comment]
Fixed profiling (skipping 1st compilation step).
Updated README with profile instructions (and unified tabs/spaces).
Tested script on a fresh VM.

=> should be good to submit.

**Review by sayakpaul — APPROVED** [review]
Very nice! Thanks :)

File: examples/controlnet/train_controlnet_flax.py [hunk 2]
```diff
@@ -1020,10 +1047,19 @@ def cumul_grad_step(grad_idx, loss_grad_rng):
         )
         # train
         for batch in train_dataloader:
+
+            if args.profile_steps and global_step == 1:
+                train_metric['loss'].block_until_ready()
```
Comment (sayakpaul) [inline examples/controlnet/train_controlnet_flax.py:1052]: Should this be `train_metrics`?

**Comment by sayakpaul** [comment]
@andsteing before merging, we need to ensure that the code quality checks pass. So, for that, I was running `make style && make quality` ([reference](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)) locally. It gave me:

```bash
ruff examples scripts src tests utils --fix
examples/controlnet/train_controlnet_flax.py:1051:17: F821 Undefined name `train_metric`
examples/controlnet/train_controlnet_flax.py:1054:17: F821 Undefined name `train_metric`
Found 2 errors.
make: *** [style] Error 1
```

Could you fix that? Possibly related: https://github.com/huggingface/diffusers/pull/3053#discussion_r1164193990

File: examples/controlnet/train_controlnet_flax.py (hunk 2 above)
Comment (andsteing) [reply examples/controlnet/train_controlnet_flax.py]: No, this is indeed `train_metric` (note that `train_metrics` is a list). Initialized it before the loop.

(I didn't notice the failing `make quality` check, my bad!)

**Comment by sayakpaul** [comment]
Thanks much!

@yiyixuxu could you give this a review one more time and merge as you see fit?
